src/main/java/com/facebook/openwifirrm/modules/ModelerUtils.java

```diff
@@ -381,4 +386,150 @@ public static Map<String, Map<String, WifiScanEntry>> getAggregatedWifiScans(
+					}
+					bssidToAggregatedStates.put(key, aggregatedStates);
+				}
+			}
+		}
+	}
+
+	/**
+	 * This method aggregates States by bssid&station key pair and radio info.
+	 * if two States of the same bssid&station match in channel, channel width and tx_power
+	 * need to be aggregated to one {@code AggregatedState}. Currently only mcs and
+	 * rssi fields are being aggregated. They are of List<Integer> type in AggregatedState,
+	 * which list all the values over the time.
+	 *
+	 * @param dataModel the data model which includes the latest recorded States.
+	 * @param obsoletionPeriodMs the maximum amount of time (in milliseconds) it
+	 *                           is worth aggregating over, starting from the
+	 *                           most recent States and working backwards in time.
+	 * 							 A State exactly {@code obsoletionPeriodMs} ms earlier
+	 * 							 than the most recent State is considered non-obsolete
+	 *                           (i.e., the "non-obsolete" window is inclusive).
+	 *                           Must be non-negative.
+	 * @param refTimeMs	the reference time were passed to make testing easier
+	 * @return  Map<String, Map<String, List<AggregatedState>>> A map from serial number to
+	 * 							a map from bssid_station String pair to a list of AggregatedState.
+	 */
+	public static Map<String, Map<String, List<AggregatedState>>> getAggregatedStates(
+		Modeler.DataModel dataModel,
+		long obsoletionPeriodMs,
+		long refTimeMs
+	) {
+		if (obsoletionPeriodMs < 0) {
+			throw new IllegalArgumentException(
+				"obsoletionPeriodMs must be non-negative."
+			);
+		}
+		Map<String, Map<String, List<AggregatedState>>> aggregatedStates =
+			new HashMap<>();
+
+		for (
+			Map.Entry<String, List<State>> deviceToStateList : dataModel.latestStates
+				.entrySet()
+		) {
+			String serialNumber = deviceToStateList.getKey();
+			List<State> states = deviceToStateList.getValue();
+
+			if (states.isEmpty()) {
+				continue;
+			}
+
+			/**
+			 * Sort in reverse chronological order. Sorting is done just in case the
+			 * States in the original list are not chronological already - although
+			 * they are inserted chronologically, perhaps latency, synchronization, etc.
+			 */
+			states.sort(
+				(state1, state2) -> -Long.compare(state1.unit.localtime, state2.unit.localtime)
+			);
+
+			Map<String, List<AggregatedState>> bssidToAggregatedStates =
+				aggregatedStates
+					.computeIfAbsent(serialNumber, k -> new HashMap<>());
+
+			for (State state : states) {
+				if (refTimeMs - state.unit.localtime > obsoletionPeriodMs) {
+					// discard obsolete entries
+					continue;
+				}
+
+				addStateToAggregation(bssidToAggregatedStates, state);
+			}
+		}
+
+		return aggregatedStates;
+	}
+
+	/** This method gets the most latest State from latestStates per device.
+	 * @param latestStates list of latest States per device.
+	 * @return Map<String, State>  a map from device String to  most latest State.
```

Comment:
delete extra space before `a map` and change `most latest` to just `latest` (latest already means most recent)